tests/integration/upgrade_tests/test_rollback.py

```diff
@@ -1,14 +1,66 @@
 #!/usr/bin/env python3
 # Copyright 2024 Canonical Ltd.
 # See LICENSE file for licensing details.
+import pathlib
+import logging
+import time
 import pytest
+import pytest_asyncio
+import shutil
+import zipfile
+import re
+import tenacity
 from pytest_operator.plugin import OpsTest
 from ..helpers import (
     deploy_cluster_components,
+    get_juju_status,
     integrate_cluster_components,
     MONGOS_APP_NAME,
 )
 
+logger = logging.getLogger(__name__)
+
+UPGRADE_TIMEOUT = 15 * 60
+
+
+async def refresh_with_juju(ops_test: OpsTest, app_name: str, channel: str) -> None:
+    refresh_cmd = f"refresh {app_name} --model {ops_test.model.info.name} --channel {channel} --switch ch:mongodb"
+    await ops_test.juju(*refresh_cmd.split())
+
+
+@pytest_asyncio.fixture
+async def local_charm(ops_test: OpsTest):
+    charm = await ops_test.build_charm(".")
+    yield charm
+
+
+@pytest_asyncio.fixture
+def faulty_upgrade_charm(local_charm, tmp_path: pathlib.Path):
+    fault_charm = tmp_path / "fault_charm.charm"
```

Comment:
mind adding a docstring here that:
1. shortly describes what this fixture does
2. the importance of it

Reply:
Sure